### PR TITLE
Port to Typescript, abstract strategies, use cookies

### PR DESCRIPTION
**Pull request by connor4312**
Bringing this back into the frontend for two issues:

 - Duplicates of notifications appearing when multiple Beam tabs are open (using the cookie strategy)
 - Duplicates of mature toasts appearing (using the memory strategy)

### CODE REVIEW

File: src/lockal.ts [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+   * @override
+   */
+  public acquire(key: string, id: string, ttl: number): Promise<void> {
+    const value = memoryStore[key];
+    const now = Date.now();
+    if (value && now < value.expiresAt) {
+      if (value.id === id) {
+        clearTimeout(value.removerTimeout);
+      } else {
+        return Promise.reject(new LockFailedError('Failed to acquire the lock'));
```
Comment (saadataz) [inline src/lockal.ts]: more specific error msg would help. Also shouldnt this be out of the if statement?

File: src/lockal.ts [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+export class MemoryStrategy implements ILockStrategy {
+  /**
+   * @override
+   */
+  public acquire(key: string, id: string, ttl: number): Promise<void> {
+    const value = memoryStore[key];
+    const now = Date.now();
+    if (value && now < value.expiresAt) {
+      if (value.id === id) {
+        clearTimeout(value.removerTimeout);
```
Comment (saadataz) [inline src/lockal.ts]: is that supposed to extend the exiting lock??

File: src/lockal.ts (hunk 1 above)
Comment (connor4312) [inline src/lockal.ts]: nope -- we want to reject only if there was a non-expired value with an ID different than our own (i.e. someone else is currently holding the lock)

File: src/lockal.ts (hunk 2 above)
Comment (connor4312) [inline src/lockal.ts]: In this case it doesn't matter too much, we overwrite it right below here with a new object

**Pull request merged**